library/BerlinPipeline/src/Juvix/BerlinPipeline/Env.hs

```diff
@@ -37,23 +39,65 @@ newtype EnvS b = EnvS (State T b)
 
 data StopADT = Stop
 
+-- | Register the pipeline function to the environment
 registerStep :: CircularList.T Step.Named -> EnvS ()
-registerStep = notImplemented
+registerStep l = do
+  modify@"registeredPipeline" $ \i -> i <> l
 
+-- | Create a named group of pipeline steps or nested grouping of pipeline steps.
 defPipelineGroup :: NameSymbol.T -> [CircularList.T Step.Named] -> CircularList.T Step.Named
-defPipelineGroup = notImplemented
+defPipelineGroup sym ls = foldl' (<>) (CircularList.init sym) ls
 
+-- | Tell the environment to stop at a particular step when running the environment.
 stopAt :: NameSymbol.T -> EnvS ()
-stopAt = notImplemented
+stopAt sym = put@"stoppingStep" (Just sym)
 
 stopAtNothing :: EnvS ()
-stopAtNothing = notImplemented
-
-eval :: T -> Pipeline.ComputationalInput
-eval = notImplemented
+stopAtNothing = put@"stoppingStep" Nothing
 
-run :: EnvS b -> T -> Pipeline.ComputationalInput
-run = notImplemented
+-- Change our environment
+eval :: MonadIO m => T -> m Pipeline.CIn
+eval
+  ( T
+      input@(Pipeline.CIn wEnv@(Pipeline.WorkingEnv sexp context) surr)
+      pipeline@(CircularList.T reclist)
+      stoppingStep
+    ) = do
+    case nextStep of
+      Nothing -> pure input
+      Just (CircularList.NonCircSchema nStep) -> do
+        let name = Step.name nStep
+        let namedInput = Pipeline.nameCIn name input
+        if shouldStop stoppingStep name
+          then pure input
+          else liftIO $ do
+            let (Step.T step) = Step.step nStep
+            res <- step namedInput
+            case res of
+              Pipeline.Success {meta, result} ->
+                eval $
+                  T
+                    { information =
+                        Pipeline.CIn
+                          { languageData = result,
+                            surroundingData = Pipeline.SurroundingEnv Nothing meta
+                          },
+                      registeredPipeline = remainder,
+                      stoppingStep
+                    }
+              Pipeline.Failure {meta, partialResult} -> pure $ case partialResult of
+                Nothing -> Pipeline.metaCIn meta namedInput
+                Just wEnv ->
+                  let out = Pipeline.metaCIn meta namedInput
+                   in out {Pipeline.languageData = wEnv}
+      Just (CircularList.CircSchema ls) -> notImplemented
+    where
+      shouldStop (Just n) named
+        | n == named = True
+        | otherwise = False
+      shouldStop _ _ = False
+      nextStep = CircularList.firstNested pipeline
+      remainder = CircularList.removeFirstNested pipeline
 
-extract :: EnvS b -> T
-extract = notImplemented
```

Comment:
We should still have an extract, even if it's in `IO` or gets passed an explicit `Context.T`. As it's a useful function to have